libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/MapboxTripStarter.kt

```diff
@@ -0,0 +1,171 @@
+        val granted = PermissionsManager.areLocationPermissionsGranted(context)
+        optionsFlow.update { it.toBuilder().isLocationPermissionGranted(granted).build() }
+
+        // Observe any changes to the options
+        coroutineScope.launch {
+            optionsFlow.collect { options ->
+                onStarterOptionsChanged(mapboxNavigation, options)
+            }
+        }
+    }
+
+    /**
+     * Signals that the [mapboxNavigation] instance is being detached.
+     * @param mapboxNavigation
+     */
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        coroutineScope.cancel()
+        onTripDisabled(mapboxNavigation)
+        this.mapboxNavigation = null
+    }
+
+    /**
+     * Get the options that are currently set.
+     */
+    fun getOptions(): MapboxTripStarterOptions = optionsFlow.value
+
+    /**
+     * Set new options.
+     */
+    fun setOptions(options: MapboxTripStarterOptions) = apply {
+        checkOptions(options)
+        this.optionsFlow.value = options
+    }
+
+    /**
+     * Apply changes to existing options.
+     */
+    fun update(function: (MapboxTripStarterOptions.Builder) -> MapboxTripStarterOptions.Builder) {
+        this.optionsFlow.update {
+            val nextOptions = function.invoke(it.toBuilder()).build()
+            checkOptions(nextOptions)
+            nextOptions
+        }
+    }
+
+    /**
+     * Throws an error if location permissions are set to true but the location permissions are
+     * not actually granted.
+     */
+    private fun checkOptions(options: MapboxTripStarterOptions) {
+        val mapboxNavigation = this.mapboxNavigation
+        checkNotNull(mapboxNavigation) {
+            "MapboxTripStarter cannot be used while MapboxNavigation is detached."
+        }
+        val context = mapboxNavigation.navigationOptions.applicationContext
+        val granted = options.isLocationPermissionGranted
+        if (granted && !PermissionsManager.areLocationPermissionsGranted(context)) {
+            error(
+                "updateLocationPermissions can only be set to true when location permissions" +
+                    " have been accepted."
+            )
+        }
+    }
+
+    private fun onStarterOptionsChanged(
+        mapboxNavigation: MapboxNavigation,
+        options: MapboxTripStarterOptions
+    ) {
+        if (options.tripType == MAPBOX_TRIP_STARTER_REPLAY_ROUTE) {
+            onReplayTripEnabled(mapboxNavigation)
+        } else if (options.tripType == MAPBOX_TRIP_STARTER_FOLLOW_DEVICE &&
+            options.isLocationPermissionGranted
+        ) {
+            onTripSessionEnabled(mapboxNavigation)
+        } else {
+            onTripDisabled(mapboxNavigation)
+        }
+    }
+
+    private fun onReplayTripEnabled(mapboxNavigation: MapboxNavigation) {
+        if (mapboxNavigation.getTripSessionState() != TripSessionState.STARTED &&
+            !mapboxNavigation.isReplayEnabled()
```

Comment:
Because `ReplayRouteSession` is not part of `MapboxNavigation` it is requiring some special logic. [The regular trip session](https://github.com/mapbox/mapbox-navigation-android/blob/6c4a1d8551ef38f205ca4181cd3fb4226f9bb224/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L322-L329) ensures that a started trip session is not restarted.

But if we call onDetached/onAttached every time the options change it will cause the replayer to restart at the beginning of the route (🤔  that may be what we want..)

Reply:
updating the location permissions still causes a reset. so there is some more logic that needs to change in order to remove this condition